### PR TITLE
Upload API wrapper

### PR DESCRIPTION
**Pull request by idonov8**

### CODE REVIEW

File: dagshub/upload/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+		elif "ACCESS_TOKEN" in os.environ:
+			self.authToken = os.environ["ACCESS_TOKEN"]
+		else:
+			raise Exception("Can't find access token. Please set enviroment variable ACCESS_TOKEN with a DagsHub access token")
+		# TODO: verify token
+		
+		if src_url != None:
+			self.src_url = src_url
+		elif "SRC_URL" in os.environ:
+			self.src_url = os.environ["SRC_URL"]
```
Comment (simonlsk) [inline dagshub/upload/wrapper.py:27]:
```suggestion
		self.src_url = src_url if src_url is not None else os.environ.get("SRC_URL")
```

File: dagshub/upload/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+		if branch != None:
```
Comment (simonlsk) [inline dagshub/upload/wrapper.py:29]:
```suggestion
		if branch is not None:
```

File: dagshub/upload/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+			# Find default branch and set it if not specified
+			res = requests.get(urllib.parse.urljoin(self.src_url, REPO_INFO_URL.format(
+				owner=self.owner,
+				reponame=self.name
+				)))
+			try:
+				self.branch = res.json().get('default_branch')
+				print("Set default branch: ", self.branch)
+			except:
+				raise Exception("Failed to get default branch for repository. Please specify a branch and make sure repository details are correct.")
```
Comment (simonlsk) [inline dagshub/upload/wrapper.py:41]: Maybe move all this to a method named `_get_default_branch`

**Review by simonlsk — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -96,6 +96,24 @@ All of the supported ways to enable DagsHub Streaming should automatically detec
...
+A basic use looks like this:
+```python
+from dagshub.upload import Repo
+
+repo = Repo("idonov8", "baby-yoda-segmentation-dataset", branch="new_annotations")
+ds = repo.directory("images")
+
+with open("test_photo.png", 'rb') as f:
+    # 'target_dir' is just the enclosing directory name. 
+    ds.add(file=f, target_dir="test_images")
```
[the remark this answers is not in the document]
Comment (idonov8) [reply README.md:110]: I feel that having only the option to input a full path will result a lot of repetition of the file name, and @guysmoilov argued that having `path` not be the full path is not intuitive. 

So I just decided to have both options 🤷‍♂️

File: dagshub/upload/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+			raise Exception("You must provide a valid commit message")
+		data["commit_message"] = self.commit_data.message
+
+		# Prints for debugging
+		print("URL: ", self.request_url)
+		print("DATA:")
+		pprint(data)
+		print("Files:")
+		pprint(self.files)
+		print("making request...")
```
[the remark this answers is not in the document]
Comment (idonov8) [reply dagshub/upload/wrapper.py:119]: Good idea. I think that for the alpha version these prints can be very useful for debugging, so I left it in.

File: dagshub/upload/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		res = requests.put(
+			self.request_url, 
+			data, 
+			files=[("files", file) for file in self.files], 
+			headers={'Authorization': 'token ' + self.repo.authToken})
```
[the remark this answers is not in the document]
Comment (idonov8) [reply dagshub/upload/wrapper.py:124]: That would require me to ask for `username` too, which is not really needed.

File: dagshub/upload/wrapper.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+class Commit:
+	choice="direct"
+	message=""
+	summary=""
+	versioning="auto"
+	new_branch=""
+	last_commit=""
+	def __init__(self):
+		return
```
[the remark this answers is not in the document]
Comment (idonov8) [reply dagshub/upload/wrapper.py:55]: Is doing it like this better?
```python
class Commit:
	def __init__(self):
		self.choice="direct"
		self.message=""
		self.summary=""
		self.versioning="auto"
		self.new_branch=""
		self.last_commit=""
```

File: dagshub/upload/wrapper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (simonlsk) [reply dagshub/upload/wrapper.py:131]: That authentication method is still there but we're aiming to remove it completely, so we can't really use it, it's also not documented in the API docs so we shouldn't use it.

File: dagshub/upload/wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonlsk) [reply dagshub/upload/wrapper.py]: It's not "better" it's just probably what you meant to do. Why do you want to to be a class anyway?

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (simonlsk) [reply README.md:114]: Anyway it's not "target dir" because the path should be a file path not a dir path.

**Review by simonlsk — CHANGES_REQUESTED** [review]

File: dagshub/upload/wrapper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (idonov8) [reply dagshub/upload/wrapper.py:131]: I'll change it to `Bearer` is that cool?

File: dagshub/upload/wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (idonov8) [reply dagshub/upload/wrapper.py]: I just found myself having a lot of `commit_<something>` (commit_choice, commit_message, etc.) and thought it should probably be a separate entity.

File: README.md [hunk 8]
```diff
@@ -106,12 +106,16 @@ A basic use looks like this:
 ```python
 from dagshub.upload import Repo
 
-repo = Repo("idonov8", "baby-yoda-segmentation-dataset", branch="new_annotations")
+# Upload a single file to a repository with one line
+Repo("idonov8", "baby-yoda-segmentation-dataset").upload("file.txt", "commit message") # Optional: versioning, new_branch, last_commit, path
+
+# Upload files to dvc folder in a repository
```
Comment (simonlsk) [inline README.md:112]: I think the point here is "upload multiple files in a directory in a single commit"

File: README.md [hunk 9]
```diff
@@ -106,12 +106,16 @@ A basic use looks like this:
 ```python
 from dagshub.upload import Repo
 
-repo = Repo("idonov8", "baby-yoda-segmentation-dataset", branch="new_annotations")
+# Upload a single file to a repository with one line
+Repo("idonov8", "baby-yoda-segmentation-dataset").upload("file.txt", "commit message") # Optional: versioning, new_branch, last_commit, path
```
Comment (simonlsk) [inline README.md:110]: What about authentication? You can't upload without authenticating

Comment (simonlsk) [reply README.md:110]: I would expect username + token / password here somewhere

**Review by simonlsk — CHANGES_REQUESTED** [review]

File: dagshub/upload/wrapper.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+		if branch is not None:
+			self.branch = branch
+		else:
+			print("Branch wasn't provided. Fetching default branch...")
+			self._set_default_branch()
+		print("Set branch: ", self.branch)
+
+	def upload(self, file: Union[str, IOBase], message, versioning=None, new_branch=None, last_commit=None, path=None):
+		ds = DataSet(self, ".")
```
Comment (simonlsk) [inline dagshub/upload/wrapper.py:41]: It's not urgent now, but I think this is semantically wrong.
If I update some random dvc or git file, I'm not updating a dataset, I'm just uploading a file. DRY is good but the code still needs to tell a story that is correct.

File: README.md (hunk 9 above)
Comment (idonov8) [reply README.md:110]: There is a section underneath. I'll remove the OAuth from there.

File: dagshub/upload/wrapper.py (hunk 10 above)
Comment (idonov8) [reply dagshub/upload/wrapper.py:41]: 🤔

**Pull request merged**